store/tikv/region_request.go

```diff
@@ -73,22 +74,28 @@ func NewRegionBatchRequestSender(cache *RegionCache, client Client) *RegionBatch
 	return &RegionBatchRequestSender{RegionRequestSender: RegionRequestSender{regionCache: cache, client: client}}
 }
 
-func (ss *RegionBatchRequestSender) sendReqToAddr(bo *Backoffer, ctxs []copTaskAndRPCContext, req *tikvrpc.Request, timout time.Duration) (resp *tikvrpc.Response, retry bool, err error) {
+func (ss *RegionBatchRequestSender) sendStreamReqToAddr(bo *Backoffer, ctxs []copTaskAndRPCContext, req *tikvrpc.Request, timout time.Duration) (resp *tikvrpc.Response, retry bool, cancel func(), err error) {
 	// use the first ctx to send request, because every ctx has same address.
-	ctx := ctxs[0].ctx
-	if e := tikvrpc.SetContext(req, ctx.Meta, ctx.Peer); e != nil {
-		return nil, false, errors.Trace(e)
+	cancel = func() {}
+	rpcCtx := ctxs[0].ctx
+	if e := tikvrpc.SetContext(req, rpcCtx.Meta, rpcCtx.Peer); e != nil {
+		return nil, false, cancel, errors.Trace(e)
+	}
+	ctx := bo.ctx
+	if rawHook := ctx.Value(RPCCancellerCtxKey{}); rawHook != nil {
+		ctx, cancel = rawHook.(*RPCCanceller).WithCancel(ctx)
 	}
-	resp, err = ss.client.SendRequest(bo.ctx, ctx.Addr, req, timout)
+	resp, err = ss.client.SendRequest(ctx, rpcCtx.Addr, req, timout)
```

Comment:
If RPCCanceller has been cancelled, do we need to send request to tiflash ?

Reply:
If RPCCanceller has been cancelled, sendRequest will immediately be returned， but it may send a request for 50% probability due to go's `select case` for multiple available channels.

emmm..we can check ctx.Done before sending for every request but check the channel is cost-operator, maybe better avoid this handle for this extreme case- -？

Reply:
Ok